module/zfs/spa.c

```diff
@@ -2318,6 +2332,209 @@ vdev_count_verify_zaps(vdev_t *vd)
 #endif
 
 /*
+ * Determine whether the activity check is required.
+ */
+static boolean_t
+spa_activity_check_required(spa_t *spa, uberblock_t *ub, nvlist_t *config)
+{
+	uint64_t state = 0;
+	uint64_t hostid = 0;
+	uint64_t tryconfig_txg = 0;
+	uint64_t tryconfig_timestamp = 0;
+	nvlist_t *nvinfo;
+
+	if (nvlist_exists(config, ZPOOL_CONFIG_LOAD_INFO)) {
+		nvinfo = fnvlist_lookup_nvlist(config, ZPOOL_CONFIG_LOAD_INFO);
+		(void) nvlist_lookup_uint64(nvinfo, ZPOOL_CONFIG_MMP_TXG,
+		    &tryconfig_txg);
+		(void) nvlist_lookup_uint64(config, ZPOOL_CONFIG_TIMESTAMP,
+		    &tryconfig_timestamp);
+	}
+
+	(void) nvlist_lookup_uint64(config, ZPOOL_CONFIG_POOL_STATE, &state);
+	(void) nvlist_lookup_uint64(config, ZPOOL_CONFIG_HOSTID, &hostid);
+
+	/*
+	 * Disable the MMP activity check - This is used by zdb which
+	 * is intended to be used on potentially active pools.
+	 */
+	if (spa->spa_import_flags & ZFS_IMPORT_SKIP_MMP)
+		return (B_FALSE);
+
+	/*
+	 * Since the hostid is likely to be different before the pivot than
+	 * it is after the pivot, and since it is hard to see how two nodes
+	 * could safely boot from the same pool anyway, we do not do MMP
+	 * for root pools.
+	 */
+	if (spa_is_root(spa))
+		return (B_FALSE);
+
+	/*
+	 * Skip the test if host which last imported the pool had MMP disabled
+	 * or was using a non-MMP aware version.  If the pool is successfully
+	 * imported, we will then respect the local setting for mmp writes.
```

Comment:
We could do that and it would provide some additional protection.  The intent was to provide a reasonable way to disable the activity check on systems which wouldn't benefit from it.  Where it would just be a unneeded delay.  That notion got extended to legacy pools to avoid changing the existing behavior.  But since this can only happen on the first import of a legacy pool I'd be OK with taking the more conservative approach and doing the activity check.